src/legacy/ui/public/agg_types/controls/components/input_list.tsx

```diff
@@ -74,120 +70,109 @@ interface InputListProps {
 
-  const onUpdate = (modelList: InputModel[]) => {
+  const updateValues = (modelList: InputModel[]) => {
     setModels(modelList);
     onChange(modelList.map(config.onChangeFn));
   };
-
   const onChangeValue = (index: number, value: string, modelName: string) => {
-    const range = models[index][modelName];
-    const { model, isInvalid } = validateValue(value);
-    range.value = value;
-    range.model = model;
-    range.isInvalid = isInvalid;
-    onUpdate(models);
-  };
-  const onDelete = (id: string) => {
-    const newArray = models.filter(model => model.id !== id);
-    onUpdate(newArray);
+    const { model, isInvalid } = validateValue(value, config);
+    updateValues(
+      models.map((range, arrayIndex) =>
+        arrayIndex === index
+          ? {
+              ...range,
+              [modelName]: {
+                value,
+                model,
+                isInvalid,
+              },
+            }
+          : range
+      )
+    );
   };
-
-  const onAdd = () => {
-    const newArray = [
+  const onDelete = (id: string) => updateValues(models.filter(model => model.id !== id));
+  const onAdd = () =>
+    updateValues([
       ...models,
       {
         id: generateId(),
-        ...config.defaultEmptyValue,
+        ...config.getModelValue(),
       } as InputModel,
-    ];
-    onUpdate(newArray);
-  };
-
-  const getUpdatedModels = (objList: InputObject[], modelList: InputModel[]) => {
-    if (!objList.length) {
-      return modelList;
-    }
-    return objList.map((item, index) => {
-      const model = modelList[index] || {
-        id: generateId(),
-        ...config.getModelValue(item),
-      };
-
-      config.validateModel(validateItem, item, model);
-
-      return model;
-    });
-  };
-
-  const validateItem = (value: string | undefined, modelObj: InputItem) => {
-    const { model, isInvalid } = validateValue(value);
-    if (value !== modelObj.model) {
-      modelObj.value = model;
-    }
-    modelObj.model = model;
-    modelObj.isInvalid = isInvalid;
-  };
-
-  const validateValue = (inputValue: string | undefined) => {
-    const result = {
-      model: inputValue || '',
-      isInvalid: false,
-    };
-    if (!inputValue) {
-      result.isInvalid = false;
-      return result;
-    }
-    try {
-      const InputObject = config.validateClass;
-      result.model = new InputObject(inputValue).toString();
-      result.isInvalid = false;
-      return result;
-    } catch (e) {
-      result.isInvalid = true;
-      return result;
-    }
-  };
+    ]);
 
-  const hasInvalidValues = (modelList: InputModel[]) => {
-    return !!modelList.find(config.hasInvalidValuesFn);
-  };
-
-  // responsible for discarding changes
   useEffect(() => {
-    setModels(getUpdatedModels(list, models));
-  }, [list]);
+    // resposible for setting up an initial value when there is no default value
+    if (!list.length) {
+      updateValues([
+        ...models,
```

Comment:
If `list` is empty, `models` will be empty as well. So I think we don't need `models` here.

Reply:
You're right! Thanks!